app/src/main/kotlin/cz/covid19cz/erouska/ui/login/LoginVM.kt

```diff
@@ -184,23 +181,49 @@ class LoginVM(
-        L.e(e)
-        if (e is FirebaseAuthInvalidCredentialsException) {
-            L.d("Error code: ${e.errorCode}")
+    private fun handleError(exception: Exception) {
+        if (exception is FirebaseAuthInvalidCredentialsException) {
+            L.d("Error code: ${exception.errorCode}")
         }
-        if (e is FirebaseAuthInvalidCredentialsException && e.errorCode == "ERROR_INVALID_PHONE_NUMBER") {
-            mutableState.postValue(EnterPhoneNumber(true))
-        } else if (e is FirebaseAuthInvalidCredentialsException && e.errorCode == "ERROR_INVALID_VERIFICATION_CODE" || e is FirebaseAuthInvalidCredentialsException) {
-            mutableState.postValue(EnterCode(true, phoneNumber))
-        } else if (e is FirebaseAuthInvalidCredentialsException && e.errorCode == "ERROR_TOO_MANY_REQUESTS" || e is FirebaseTooManyRequestsException) {
-            mutableState.postValue(LoginError(R.string.login_too_many_attempts_error.toText()))
-        } else if (e is FirebaseAuthInvalidCredentialsException && e.errorCode == "ERROR_SESSION_EXPIRED") {
-            mutableState.postValue(LoginError(R.string.login_session_expired.toText()))
-        } else if (e is FirebaseNetworkException) {
-            mutableState.postValue(LoginError(R.string.login_network_error.toText()))
-        } else {
-            mutableState.postValue(LoginError(R.string.unexpected_error_text.toText()))
+        when (exception) {
+            is FirebaseAuthInvalidCredentialsException -> {
+                exception.handle()
+            }
+            is FirebaseTooManyRequestsException -> {
+                mutableState.postValue(LoginError(R.string.login_too_many_attempts_error.toText()))
+            }
+            is FirebaseNetworkException -> {
+                mutableState.postValue(LoginError(R.string.login_network_error.toText()))
+            }
+            is FirebaseAuthUserCollisionException -> {
+                mutableState.postValue(LoginError(
+                    R.string.login_number_already_in_use_error.toText(phoneNumber.formatPhoneNumber())
+                ))
+            }
+            else -> {
+                mutableState.postValue(LoginError(R.string.unexpected_error_text.toText()))
+            }
+        }
+    }
+
+    private fun FirebaseAuthInvalidCredentialsException.handle() {
+        when (errorCode) {
+            "ERROR_INVALID_PHONE_NUMBER" -> {
+                mutableState.postValue(EnterPhoneNumber(true))
+            }
+            "ERROR_INVALID_VERIFICATION_CODE" -> {
+                mutableState.postValue(EnterCode(true, phoneNumber))
+            }
+            "ERROR_TOO_MANY_REQUESTS" -> {
+                mutableState.postValue(LoginError(R.string.login_too_many_attempts_error.toText()))
+            }
+            "ERROR_SESSION_EXPIRED" -> {
+                mutableState.postValue(LoginError(R.string.login_session_expired.toText()))
+            }
+            else -> {
+                L.e(this)
```

Comment:
We need to differentiate logging from the main else. Maybe wrapping it into another exception, where error code will be part of it?

Reply:
The error code should be there, or not?

Reply:
Closing the loop. It's there, but we can't see it in crashlytics 🙂 will update